xds/internal/xdsclient/tlscreds/bundle_test.go

```diff
@@ -0,0 +1,256 @@
+	}
+
+	// Leave time for the file_watcher provider to reload the CA.
+	time.Sleep(100 * time.Millisecond)
+
+	_, stopFunc := startServer(t, srvAddr.String(), tls.NoClientCert)
+	defer stopFunc()
+
+	// Client handshake should fail because the server cert is signed by an
+	// unknown CA.
+	_, err = client.EmptyCall(context.Background(), &testpb.Empty{})
+	if st, ok := status.FromError(err); !ok || st.Code() != codes.Unavailable {
+		t.Errorf("Expected unavailable error, got %v", err)
+	} else if want := "certificate signed by unknown authority"; !strings.Contains(st.Message(), want) {
+		t.Errorf("Expected call error to contain '%s', got %v", want, st.Message())
+	}
+}
+
+func TestMTLS(t *testing.T) {
+	srvAddr, stopFunc := startServer(t, "localhost:0", tls.RequireAndVerifyClientCert)
+	defer stopFunc()
+
+	cfg := fmt.Sprintf(`{
+		"ca_certificate_file": "%s",
+		"certificate_file": "%s",
+		"private_key_file": "%s"
+	}`,
+		testdata.Path("x509/server_ca_cert.pem"),
+		testdata.Path("x509/client1_cert.pem"),
+		testdata.Path("x509/client1_key.pem"))
+	tlsBundle, err := NewBundle([]byte(cfg))
+	if err != nil {
+		t.Fatalf("Failed to create TLS bundle: %v", err)
+	}
+	dialOpts := []grpc.DialOption{
+		grpc.WithCredentialsBundle(tlsBundle),
+		grpc.WithAuthority("x.test.example.com"),
+	}
+
+	t.Run("ValidClientCert", func(t *testing.T) {
+		conn, err := grpc.Dial(srvAddr.String(), dialOpts...)
+		if err != nil {
+			t.Fatalf("Error dialing: %v", err)
+		}
+		client := testgrpc.NewTestServiceClient(conn)
+
+		_, err = client.EmptyCall(context.Background(), &testpb.Empty{})
+		if err != nil {
+			t.Errorf("Error calling EmptyCall: %v", err)
+		}
+		conn.Close()
+	})
+
+	t.Run("Provider failing", func(t *testing.T) {
+		// Check that if the provider returns an errors, we fail the handshake.
+		// It's not easy to trigger this condition, so we rely on closing the
+		// provider.
+		creds, ok := tlsBundle.TransportCredentials().(*reloadingCreds)
+
+		// Force the provider to be initialized. The test is flaky otherwise,
+		// since close may be a noop.
+		_, _ = creds.provider.KeyMaterial(context.Background())
+
+		if !ok {
+			t.Fatalf("Expected reloadingCreds, got %T", tlsBundle.TransportCredentials())
+		}
+
+		creds.provider.Close()
+
+		conn, err := grpc.Dial(srvAddr.String(), dialOpts...)
+		if err != nil {
+			t.Fatalf("Error dialing: %v", err)
+		}
+		client := testgrpc.NewTestServiceClient(conn)
+		_, err = client.EmptyCall(context.Background(), &testpb.Empty{})
+		if st, ok := status.FromError(err); !ok || st.Code() != codes.Unavailable {
+			t.Errorf("Expected unavailable error, got %v", err)
+		} else if want := "provider instance is closed"; !strings.Contains(st.Message(), want) {
+			t.Errorf("Expected error to contain '%s', got %v", want, st.Message())
+		}
+		conn.Close()
+	})
+}
+
+type stopFunc func()
+
+func startServer(t *testing.T, addr string, clientAuth tls.ClientAuthType) (net.Addr, stopFunc) {
```

Comment:
You could end up using existing test utilities that we have:
- spin up a server using the `stubserver` package. It implements the test service for you and allows you to pass server options and dial options (if you want it to create the clientConn as well).
- To setup cert files: https://github.com/grpc/grpc-go/blob/df02c114bbc8c8580b2e96a125ab1d0a774b8f16/internal/testutils/xds/e2e/setup_certs.go#L77
- To setup cert files to be used by the server, I've added the same as part of an ongoing PR. See: https://github.com/grpc/grpc-go/pull/6848/files#diff-2647d1d4b435368cb2ddc0827563a91a4269a32f47d721d907a2bc9cb26b55d0

Reply:
Sounds good.